lib/Dialect/FIRRTL/FIRRTLDialect.cpp

```diff
@@ -277,6 +293,18 @@ void FIRRTLDialect::printType(Type type, DialectAsmPrinter &os) const {
 Operation *FIRRTLDialect::materializeConstant(OpBuilder &builder,
                                               Attribute value, Type type,
                                               Location loc) {
+
+  // Boolean constants. Boolean attributes are always a special constant type
+  // like ClockType and ResetType.  Since BoolAttrs are also IntegerAttrs, its
+  // important that this goes first.
+  if (auto attrValue = value.dyn_cast<BoolAttr>()) {
+    auto isSpecialConstantType =
+        type.isa<ClockType, AsyncResetType, ResetType>();
+    assert(isSpecialConstantType &&
+           "BoolAttrs can only be materialized for special constant types.");
```

Comment:
Cool, thanks for adding this.  One nit is that this will cause a "warning: isSpecialConstantType is unused" in non-assert builds.  Please add a `(void)isSpecialConstantType;` or (better yet) inline the predicate to avoid this.